### PR TITLE
Align dev dependencies and linting with upstream

### PR DESCRIPTION
**Pull request by jakob-keller**
### Description of Change
Dev dependencies and linting deviate significantly between `botocore` (upstream) and `aiobotocore`. This PR works towards alignment by:
* leveraging `pre-commit` to orchestrate linting and bump hooks
* dropping dev dependencies covered by `pre-commit`
* bumping remaining dev dependencies
* including `setuptools` as dev dependency for Python >=3.12
* leveraging `pyproject.toml` to configure `pytest`
* leveraging `setup.cfg` to configure `flake8`

### CODE REVIEW

File: requirements-dev.in [hunk 1]
```diff
@@ -1,16 +1,16 @@
...
-coverage~=7.2.7
-flake8~=3.9.0
+coverage~=7.3.2
+flake8~=6.1.0
 flake8-black~=0.3.3
-flake8-isort~= 4.1.1
-black~=22.6.0
-isort~= 5.10.1
+flake8-isort~= 6.1.0
+black~=23.10.0
```
Comment (thehesiod) [inline requirements-dev.in:6]: we should probably do this like botocore and match their version, see: https://github.com/boto/botocore/blob/develop/.pre-commit-config.yaml#L37

Comment (jakob-keller) [reply requirements-dev.in:6]: I will look into that!

Comment (jakob-keller) [reply requirements-dev.in:6]: I added pre-commit as a dev dependency, aligned the pre-commit config with upstream and did a rebase. Do you want me to remove flake8, black and isort from dev dependencies and point everyone to pre-commit?

**Comment by thehesiod** [comment]
wow, u da man

**Pull request merged**